README.md
## CBNplot: Bayesian network plot for the enrichment analysis results

  <!-- badges: start -->
  [![R-CMD-check](https://github.com/noriakis/CBNplot/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/noriakis/CBNplot/actions/workflows/R-CMD-check.yaml)
  <!-- badges: end -->
 
Plot bayesian network inferred from expression data based on the enrichment analysis results from libraries including `clusterProfiler` and `ReactomePA` using `bnlearn`.

### Installation

```R
BiocManager::install("CBNplot")
```

```R
library(devtools)
install_github("noriakis/CBNplot")
```

### Usage
- Documentation: [https://noriakis.github.io/software/CBNplot/](https://noriakis.github.io/software/CBNplot/)
- Web server: [https://cbnplot.com](https://cbnplot.com)

### Plot examples

- The gene-to-gene relationship compared with the reference network.
<img src="https://github.com/noriakis/software/blob/main/images/CBNplot_readme_1.png?raw=true" width="800px">

- The plot is customizable highliting edges and nodes like hub genes.
<img src="https://github.com/noriakis/software/blob/main/images/CBNplot_readme_2.png?raw=true" width="800px">

- The example using `MicrobiomeProfiler`, thanks to the fix by [@xiangpin](https://github.com/xiangpin/).
``` R
library(MicrobiomeProfiler)
data(Rat_data)
ko.res <- enrichKO(Rat_data)
exp.dat <- matrix(abs(rnorm(910)), 91, 10) %>% magrittr::set_rownames(value=Rat_data) %>% magrittr::set_colnames(value=paste0('S', seq_len(ncol(.))))
exp.dat %>% head
library(CBNplot)
bngeneplot(ko.res, exp=exp.dat, pathNum=1, orgDb=NULL)
```
<img src="https://github.com/noriakis/software/blob/main/images/CBNplot_readme_MP.png?raw=true" width="800px">

- Another customized plot.
<img src="https://github.com/noriakis/software/blob/main/images/CBNplot_readme_3.png?raw=true" width="800px">


NEWS
Changes in version 0.99.0 (2021-07-29)
+ Prepared for submission to Bioconductor

Changes in version 0.99.1 (2022-02-24)
+ Fixed errors and warnings raised in pre-checking

Changes in version 0.99.3 (2023-04-07)
+ Fixed the problems raised in the review

Changes in version 0.99.4 (2023-04-07)
+ Fixed the WARNINGS in package builder at bioconductor.org

Changes in version 0.99.5 (2023-04-13)
+ Changed the dependency for R version


tests/testthat.R
library(testthat)
library(CBNplot)

test_check("CBNplot")


tests/testthat/test-utils.R
test_that("utilities: not produce errors in variou parameters", {
    data("exampleEaRes");data("exampleGeneExp");
    library(bnlearn)
    data("exampleEaRes")
    data("exampleGeneExp")
    net <- bngeneplot(exampleEaRes, exampleGeneExp, R=5,
                   pathNum=1, returnNet=TRUE)
    fitted <- bn.fit(net$av, net$df)
    res <- queryCpDistLw(fitted, candidate="ERCC4", evidence="ERCC2",
                          levels=c(0.1, 0.5, 0.8), n=50, alpha=TRUE)
    res <- queryCpDistLw(fitted, candidate="ERCC4", evidence="ERCC2",
                          levels=c(0.1, 0.5, 0.8), n=50, alpha=FALSE)
    res <- queryCpDistLw(fitted, candidate="ERCC4", evidence="ERCC2",
                          levels=c(0.1, 0.5, 0.8), n=50, point=TRUE)
    res <- queryCpDistLw(fitted, candidate="ERCC4", evidence="ERCC2",
                          levels=c(0.1, 0.5, 0.8), n=50, point=FALSE)
})


tests/testthat/test-bngeneplot.R
test_that("bngeneplot: not produce errors in basic usages of variou parameters", {
    data("exampleEaRes");data("exampleGeneExp");
    R <- 5
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, edgeLink=TRUE), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, hub=2), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, convertSymbol=FALSE), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, returnNet=T), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, onlyDf=T), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, showDir=T), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, strType="ms"), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1:2, R = R), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, strengthPlot=TRUE), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, delZeroDegree=FALSE), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, disc=T), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R, compareRef=T), NA)
    expect_error( bngeneplot(exampleEaRes, exampleGeneExp, pathNum = 1, R = R,
        sizeDep=T, dep=depmap::depmap_crispr()), NA)

    ## Custom
    expect_error( bngeneplotCustom(exampleEaRes, exampleGeneExp,
        pathNum = 2, R = R, glowEdgeNum=3, hub=3), NA)
    
})


R/utilitiesWeb.R

#' bootReasonOneDiscrete
#' 
#' for shiny, perform bootstrap reasoning for event of one discrete node if network is not DAG.
#' Logic sampling is used.
#'
#' @param df expression data
#' @param R bootstrap number
#' @param node candidate node (one only)
#' @param evidences expressions of evidence e.g. c("`A` < 3", "`B` >= 2")
#' @return mean and sd of frequency of event level
#' @importFrom dplyr summarise_all mutate ungroup
#' @importFrom bnlearn as.bn
#' @noRd
bootReasonOneDiscrete <- function (df, R, node, evidences, algo="hc", ref=NULL, n=NULL, algorithm.args=NULL, cont=NULL) {
    bReason <- list()
    bDif <- list()
    
    for (i in seq_len(R)) {
        sampled <- sample(nrow(df), nrow(df), replace = TRUE)
        struc <- eval(parse(text=paste0(algo,"(df[sampled,]", algorithm.args,")")))

        if (!is.null(ref)){
            struc <- as.bn(igraph::intersection(bnlearn::as.igraph(ref),bnlearn::as.igraph(struc)))
            if (!igraph::is.dag(as.igraph(struc))){
                print("intersection not DAG")
                break
            } else {
                print("intersection DAG")
            }
        }

        fitted <- bn.fit(struc, df)
        cpdRes <- queryCpDistLsWeb(fitted, candidate=node, evidences=evidences, n=n)
        cpdRes <- cpdRes$df
        cpdRes <- cpdRes %>% group_by(level, !!!syms(node)) %>%
                  summarise(n = n()) %>%
                  mutate(freq = n / sum(n))

        bReason[[i]] <- cpdRes
    }

    wmMean <- bReason %>%
            purrr::reduce(inner_join, by = c("level", node))
    wmMean <- cbind(
        wmMean %>% ungroup() %>% dplyr::select(1,2),
        wmMean %>% ungroup() %>% dplyr::select(starts_with("freq")) %>% summarize(freqMean= rowMeans(., na.rm=TRUE), freqStdev=rowSds(as.matrix(.), na.rm=TRUE)),
        wmMean %>% ungroup() %>% dplyr::select(starts_with("n")) %>% summarize(nMean= rowMeans(., na.rm=TRUE), nStdev=rowSds(as.matrix(.), na.rm=TRUE))
    )
    return(wmMean)
}


#' bootReasonOneWeb
#' 
#' for shiny, if event is numeric and the network is not DAG,
#' produce event distribution for one node from list of multiple evidences.
#'
#' @param df expression data
#' @param R bootstrap number
#' @param node candidate node (one only)
#' @param evidences list of evidences e.g. list(A=2, B=3)
#' @return mean and sd of weighted mean
#' 
#' @noRd
bootReasonOneWeb <- function (df, R, node, evidences, algo="hc", ref=NULL, n=NULL, algorithm.args=NULL, cont=NULL) {
    bReason <- list()
    bDif <- list()
    
    for (i in seq_len(R)) {
        sampled <- sample(nrow(df), nrow(df), replace = TRUE)
        struc <- eval(parse(text=paste0(algo,"(df[sampled,]", algorithm.args,")")))
        if (!is.null(ref)){
            struc <- as.bn(igraph::intersection(bnlearn::as.igraph(ref),bnlearn::as.igraph(struc)))
            if (!igraph::is.dag(as.igraph(struc))){
                print("intersection not DAG")
                break
            } else {
                print("intersection DAG")
            }
        }
        fitted <- bn.fit(struc, df)
        cpdRes <- queryCpDistLwWeb(fitted, candidate=node, evidenceList=evidences, n=n)
        cpdRes <- cpdRes$df
        tmpWm <- cpdRes %>% group_by(level) %>% summarize_at(node, ~ weighted.mean(., weights))
        bReason[[i]] <- tmpWm
        
        if (!is.null(cont)){
            difTbl <- c()
            cntTib <- tmpWm %>% filter(level==cont)
            for (lvl in level[!level %in% cont]) {
                lvlTib <- tmpWm %>% filter(level==lvl)
                difTib <- lvlTib[,node] - cntTib[,node]
                difTbl <- rbind(difTbl, difTib)
            }
            difTbl$level <- level[!level %in% cont]
            bDif[[i]] <- difTbl
        }
    }
    wmMean <- bReason %>%
        purrr::reduce(inner_join, by = "level")
    
    if (!is.null(cont)){
        wmDif <- bDif %>%
            purrr::reduce(inner_join, by = "level")
    }
    
    res <- c()
    difres <- c()
    
    for (n in node) {
        if (!is.null(cont)){
            difMean <- wmDif %>%
                select(starts_with(n)) %>%
                select(-1) %>% 
                summarize(Mean= rowMeans(., na.rm=TRUE), stdev=rowSds(as.matrix(.), na.rm=TRUE))
            difMean$level <- level[!level %in% cont]
            difMean$node <- n
            difres <- rbind(difres, difMean)
        }
        
      currentLevel <- wmMean$level
      levelMean <- wmMean %>%
        select(starts_with(n)) %>%
        select(-1) %>% 
        summarize(Mean= rowMeans(., na.rm=TRUE), stdev=rowSds(as.matrix(.), na.rm=TRUE))
      levelMean$level <- currentLevel
      levelMean$node <- n
      res <- rbind(res, levelMean)
    }
    resList <- list()
    resList[["distMean"]] <- res
    if (!is.null(cont)){
        resList[["control"]] <- paste(evidence,"=",cont)
        resList[["difMean"]] <- difres
    }
    return(resList)
}


#' queryCpDistLsWeb
#'
#' for shiny, perform cpdist and produce plot if event is factor and network is DAG
#'
#' @param fitted bn.fit object
#' @param candidate One event node
#' @param evidences expressions of evidence e.g. c("`A` < 3", "`B` >= 2")
#' @param discPalette palette to be used in scale_fill_brewer
#'
#' @noRd
queryCpDistLsWeb <- function (fitted, candidate, evidences, discPalette="Set2",...) {
    conc <- c()
    for (e in evidences){
        tmp <- eval(parse(text=paste0("cpdist(fitted, nodes='",candidate,"', evidence=",e,", method='ls', ...)")))
        conc <- rbind(conc, data.frame(tmp, rep(e, nrow(tmp))))
    }
    colnames(conc) <- c(candidate, "level")
    returnList <- list()
    returnList[["df"]] <- conc
    groupedDf <- conc %>% group_by(level, !!!syms(candidate)) %>%
                  summarise(n = n()) %>%
                  mutate(freq = n / sum(n))
    groupedPlot <- groupedDf %>%  ggplot(aes(level,freq,fill=eval(parse(text=candidate))))+
              geom_bar(stat="identity",position='dodge')+
              theme_bw()+scale_fill_brewer(palette=discPalette)

    returnList[["grouped"]] <- groupedDf
    returnList[["plot"]] <- groupedPlot
    return(returnList)
}


#' queryCpDistLwWeb
#'
#' for shiny, perform cpdist and produce plot if event is numeric and network is DAG
#'
#' @param fitted bn.fit object
#' @param candidate ONE event node
#' @param evidenceList list of evidences e.g. list(A=2, B=3)
#' @param point whether to show point of weighted mean on plot
#' @param pointSize point size
#' @param alpha The weight is shown by alpha (TRUE) or color (FALSE)
#'
#' @noRd
queryCpDistLwWeb <- function (fitted, candidate, evidenceList, point=FALSE, pointSize=5, alpha=TRUE, ...) {
    conc <- c() # Store the data per level
    for (l in seq_len(length(evidenceList))){
        ## Make label according to list
        labeller <- paste(names(evidenceList[[l]]), "=", evidenceList[[l]], collapse="_")
        tmp <- cpdist(fitted, nodes=c(candidate), evidence=evidenceList[[l]], method="lw", debug=FALSE, ...)
        conc <- rbind(conc, data.frame(tmp, attributes(tmp)$weights, rep(labeller, nrow(tmp))))
    }
    colnames(conc) <- c(candidate, "weights", "level")
    conc$level <- factor(conc$level)
    wm <- conc %>% group_by(level) %>% dplyr::summarise_at(candidate, ~ weighted.mean(., weights))
    returnList <- list()
    returnList[["df"]] <- conc
    returnList[["wm"]] <- wm
    if (length(candidate)==1){
        ## Points are colored | set alpha by their weights
        if (alpha){
            p <- conc %>% ggplot()+
                  ggdist::geom_dots(aes(y=level, x=get(candidate), fill=level, alpha=weights), color=NA)+
                  theme_bw()+xlab(candidate)
            if (point) {p <- p + geom_point(wm, mapping=aes(x=weightedMean, y=level, fill=level), shape=21, size=pointSize) }
        } else {
            p <- conc %>% ggplot()+
                  ggdist::geom_dots(aes(y=level, x=get(candidate), fill=weights), color=NA)+
                  scale_fill_ramp_continuous()+
                  theme_bw()+xlab(candidate)
            if (point) {p <- p + geom_point(wm, mapping=aes(x=weightedMean, y=level), color="tomato", size=pointSize) }
        }
        p 
        returnList[["plot"]] <- p
    }
    return(returnList)
}


#' bootReasonMultipleWeb
#' 
#' for shiny, reflect difference in color of plot when the network is not dag
#'
#' @param df expression data
#' @param R bootstrap number
#' @param nodes candidate node (accept multiple)
#' @param evidences list of evidences e.g. list(A=2, B=3)
#' @return mean and sd of weighted mean
#' @noRd
bootReasonMultipleWeb <- function (df, R, nodes, evidences, algo="hc", ref=NULL, n=NULL, algorithm.args=NULL) {
    mm <- list()
    bReason <- list()
    for (i in seq_len(R)) {
        sampled <- sample(nrow(df), nrow(df), replace = TRUE)
        struc <- eval(parse(text=paste0(algo,"(df[sampled,]", algorithm.args,")")))

        if (!is.null(ref)){
            struc <- as.bn(igraph::intersection(bnlearn::as.igraph(ref),bnlearn::as.igraph(struc)))
            if (!igraph::is.dag(as.igraph(struc))){
                print("intersection not DAG")
                break
            } else {
                print("intersection DAG")
                mm[[i]] <- struc
            }
        }

        fitted <- bn.fit(struc, df)
        
        labeller <- paste(names(evidences), "=", evidences, collapse="_")
        cpdRes <- cpdist(fitted, nodes=nodes, evidence=evidences, method="lw", n=n)
        tmpWm <- cpdRes %>% summarise_all(function(x) weighted.mean(x, attributes(cpdRes)$weights))
        tmpWm$level <- labeller
        bReason[[i]] <- tmpWm
    }

    wmMean <- bReason %>%
        purrr::reduce(rbind) %>% group_by(level) %>% summarise_all(list(mean, sd))
    
    return(list(wmMean,mm))
}


R/bngeneplotCustom.R
#' bngeneplotCustom
#'
#' Plot gene relationship within the specified pathway using customized theme
#'
#' @param results the enrichment analysis result
#' @param exp gene expression matrix
#' @param expRow the type of the identifier of rows of expression matrix
#' @param expSample candidate rows to be included in the inference
#'                  default to all
#' @param algo structure learning method used in boot.strength()
#'             default to "hc"
#' @param algorithm.args parameters to pass to 
#'                       bnlearn structure learnng function
#' @param otherVar other variables to be included in the inference
#' @param otherVarName the names of other variables
#' @param R the number of bootstrap
#' @param onlyDf return only data.frame used for inference
#' @param pathNum the pathway number
#'                (the number of row of the original result, ordered by p-value)
#' @param convertSymbol whether the label of resulting network is
#'                      converted to symbol, default to TRUE
#' @param bypassConverting bypass the symbol converting
#'                         ID of rownames and those listed in EA result
#'                         must be same
#' @param cexCategory scaling factor of size of nodes
#' @param cl cluster object from parallel::makeCluster()
#' @param showDir show the confidence of direction of edges
#' @param chooseDir if undirected edges are present, choose direction of edges
#' @param labelSize the size of label of the nodes
#' @param layout ggraph layout, default to "nicely"
#' @param strType "normal" or "ms" for multiscale implementation
#' @param dep the tibble storing dependency score from library depmap
#' @param sizeDep whether to reflect DepMap score to the node size
#' @param cellLineName the cell line name to be included
#' @param strengthPlot append the barplot depicting edges with high strength
#' @param nStrength specify how many edges are included in the strength plot
#' @param strThresh the threshold for strength
#' @param hub visualize the genes with top-n hub scores
#' @param fontFamily font family name to be used for plotting
#' @param glowEdgeNum edges with top-n confidence of direction are highlighted
#' @param nodePal vector of coloring of nodes (low, high)
#' @param edgePal vector of coloring of edges (low, high)
#' @param textCol color of texts in network plot
#' @param titleCol color of title in network plot
#' @param backCol color of background in network plot
#' @param barTextCol text color in barplot
#' @param barPal bar color
#' @param edgeLink use geom_edge_link() instead of geom_edge_diagonal()
#' @param barBackCol background color in barplot
#' @param barLegendKeyCol legend key color in barplot
#' @param barAxisCol axis color in barplot
#' @param barPanelGridCol panel grid color in barplot
#' @param returnNet whether to return the network
#' @param titleSize the size of title
#' @param scoreType score type to use on inference
#' @param orgDb perform clusterProfiler::setReadable
#'                based on this organism database
#' @param interactive whether to use bnviewer (default to FALSE)
#' @param disc discretize the expressoin data
#' @param tr Specify data.frame if one needs to discretize
#'           as the same parameters as the other dataset
#' @param remainCont Specify characters when perform discretization,
#'                   if some columns are to be remain continuous
#' @param seed A random seed to make the analysis reproducible, default is 1.
#' @param bg.colour parameter to pass to geom_node_text
#' @param bg.r parameter to pass to geom_node_text
#' 
#' @examples
#' data("exampleEaRes");data("exampleGeneExp")
#' res <- bngeneplotCustom(results=exampleEaRes, exp=exampleGeneExp,
#'                         pathNum=1, glowEdgeNum=NULL, hub=3, R=40,
#'                         fontFamily="sans")
#' @return ggplot2 object
#'
#' @export
#'

bngeneplotCustom <- function (results, exp, expSample=NULL, algo="hc", R=20,
                            pathNum=NULL, convertSymbol=TRUE, expRow="ENSEMBL",
                            interactive=FALSE, cexCategory=1, cl=NULL,
                            showDir=FALSE, chooseDir=FALSE, algorithm.args=NULL,
                            labelSize=4, layout="nicely", strType="normal",
                            returnNet=FALSE, otherVar=NULL, otherVarName=NULL,
                            onlyDf=FALSE, disc=FALSE, tr=NULL, remainCont=NULL,
                            dep=NULL, sizeDep=FALSE, orgDb=org.Hs.eg.db,
                            bypassConverting=FALSE, edgeLink=FALSE,
                            cellLineName="5637_URINARY_TRACT",
                            fontFamily="sans", strengthPlot=FALSE,
                            nStrength=10, strThresh=NULL, hub=NULL,
                            glowEdgeNum=NULL, nodePal=c("blue","red"),
                            edgePal=c("blue","red"), textCol="black",
                            titleCol="black", backCol="white",
                            barTextCol="black", barPal=c("red","blue"),
                            barBackCol="white", scoreType="bic-g",
                            barLegendKeyCol="white", barAxisCol="black",
                            bg.colour=NULL, bg.r=0.1,
                            barPanelGridCol="black", titleSize=24, seed = 1) {
    if (length(nodePal)!=2 ||
        length(edgePal)!=2 ||
        length(barPal)!=2){
            stop("Please pick two colors for nodePal, edgePal and barPal.")}
    if (!is.numeric(pathNum)){
        stop("Please specify number(s) for pathNum.")}
    if (is.null(expSample)) {
        expSample <- colnames(exp)}

    ## The newer version of reactome.db
    attributes(results)$result$Description <- gsub("Homo sapiens\r: ",
                                    "",
                                    attributes(results)$result$Description)
    
    ## Deprecated
    # if (results@keytype == "kegg"){
    #     resultsGeneType <- "ENTREZID"
    # } else {
    #     resultsGeneType <- results@keytype
    # }
    if (!bypassConverting) {
        if (!is.null(orgDb)){
            results <- setReadable(results, OrgDb=orgDb)
        }
    }
    tmpCol <- colnames(attributes(results)$result)
    tmpCol[tmpCol=="core_enrichment"] <- "geneID"
    tmpCol[tmpCol=="qvalues"] <- "qvalue"
    tmpCol[tmpCol=="setSize"] <- "Count"
    colnames(attributes(results)$result) <- tmpCol

    # if (showLineage) {
    #     if (is.null(dep)){dep = depmap::depmap_crispr()}
    #     if (is.null(depMeta)){depMeta = depmap::depmap_metadata()}
    # }

    if (sizeDep) {
        if (is.null(cellLineName)){stop("Please specify cell line name.")}
        if (is.null(dep)){dep <- depmap::depmap_crispr()}
        filteredDep <- dep %>% filter(.data$cell_line==cellLineName)
    }
    if (sizeDep){
        ## when reflecting DepMap scores to node size
        scaleSizeLow <- 1
        scaleSizeHigh <- 10
    } else {
        scaleSizeLow <- 3
        scaleSizeHigh <- 8
    }

    res <- attributes(results)$result

    genesInPathway <- unlist(strsplit(res[pathNum, ]$geneID, "/"))
    if (!bypassConverting) {
        if (!is.null(orgDb)){
            genesInPathway <- clusterProfiler::bitr(genesInPathway,
                                                    fromType="SYMBOL",
                                                    toType=expRow,
                                                    OrgDb=orgDb)[expRow][,1]
        }
    }

    pcs <- exp[ intersect(rownames(exp), genesInPathway), expSample ]

    if (!bypassConverting) {
        if (convertSymbol && !is.null(orgDb)) {
        matchTable <- clusterProfiler::bitr(rownames(pcs), fromType=expRow,
            toType="SYMBOL", OrgDb=orgDb)
        if (sum(duplicated(matchTable[,1])) >= 1) {
            message("Removing expRow that matches the multiple symbols")
            matchTable <- matchTable[
            !matchTable[,1] %in% matchTable[,1][duplicated(matchTable[,1])],]
        }
            rnSym <- matchTable["SYMBOL"][,1]
            rnExp <- matchTable[expRow][,1]
            pcs <- pcs[rnExp, ]
            rownames(pcs) <- rnSym
        }
    }

    pcs <- data.frame(t(pcs))

    if (sizeDep){
        pcs <- pcs[,intersect(filteredDep$gene_name, colnames(pcs)),]
    }
    
    geneNames <- colnames(pcs)
    
    ## Insert other vars
    if (!is.null(otherVar)) {
        pcs <- cbind(pcs, otherVar)
        if (!is.null(otherVarName)){
            colnames(pcs) <- c(geneNames, otherVarName)
        }
    }

    if (disc){
        pcsRaw <- pcs
        pcs <- discDF(pcs, tr=tr, remainCont=remainCont)
    } else {
        pcsRaw <- pcs ## Hold pcsRaw anyway
    }
    
    if (onlyDf){
        return(pcs)
    }

    if (strType == "normal"){
        strength <- withr::with_seed(seed = seed,
            boot.strength(pcs, algorithm=algo,
            algorithm.args=algorithm.args, R=R, cluster=cl))
    } else if (strType == "ms"){
        strength <- withr::with_seed(seed = seed,
            inferMS(pcs, algo=algo, algorithm.args=algorithm.args, R=R, cl=cl))
    }

    if (strengthPlot){
        strengthTop <- strength[order(strength$strength+strength$direction,
            decreasing = TRUE),][seq_len(nStrength),]
        strengthTop$label <- paste(strengthTop$from, "to", strengthTop$to)
        stp <- strengthTop %>% 
            tidyr::pivot_longer(cols=c(.data$strength, .data$direction)) %>%
            ggplot(aes(x=.data$label, y=.data$value, fill=.data$name))+
            geom_bar(position="dodge",stat="identity",alpha=0.7)+
            xlab("edges")+
            coord_flip(ylim=c(min(strengthTop$strength,
                strengthTop$direction)-0.05,1.0))+
            scale_fill_manual(values=c(barPal[1],
                barPal[2]), na.value="transparent")+
            scale_x_discrete(labels = function(x) stringr::str_wrap(x,
                                                                width = 25))+
            theme(
                text=element_text(size=16, family=fontFamily, color=barTextCol),
                plot.background = element_rect(fill = barBackCol, colour = NA),
                legend.background = ggplot2::element_blank(),
                legend.key = element_rect(fill = barLegendKeyCol),
                axis.text=element_text(color=barAxisCol),
                axis.ticks = ggplot2::element_blank(),
                axis.line = ggplot2::element_blank(),
                axis.title.y = element_text(vjust=-0.5),
                panel.grid.minor = ggplot2::element_blank(),
                panel.background = element_blank(),
                panel.grid = ggplot2::element_line(linetype = 3,
                    color = barPanelGridCol, size = 0.2))
    }

    if (!is.null(strThresh)){
        av <- averaged.network(strength, threshold=strThresh)
    } else {
        av <- averaged.network(strength)
    }

    # if (chooseDir){
    #     av <- chooseEdgeDir(av, pcs, scoreType)
    # }

    av <- cextend(av, strict=FALSE)

    g <- bnlearn::as.igraph(av)
    e <- as_edgelist(g, names = TRUE)
    eName <-paste0(e[,1], "_", e[,2])
    colnames(e) <- c("from","to")
    eDf <- merge(e, strength)
    rownames(eDf) <- paste0(eDf$from, "_", eDf$to)
    eDf <- eDf[eName, ]
    g <- set.edge.attribute(g, "color", index=E(g), eDf$strength)
    if (showDir){
        g <- set.edge.attribute(g, "label", index=E(g), round(eDf$direction,2))
    } else {
        g <- set.edge.attribute(g, "label", index=E(g), NA)
    }

    ## Define hub genes
    hScore <- hub.score(g, scale = TRUE, weights = E(g)$color)$vector

    ## Make color for glowing edges
    if (!is.null(glowEdgeNum)){
        highEdge <- E(g)[order(E(g)$label,
            decreasing=TRUE)][seq_len(glowEdgeNum)]
        glowEdge <- E(g) %in% highEdge
        E(g)$glowEdge <- E(g)$color
        E(g)$glowEdge[!glowEdge] <- NA
        E(g)$alphaEdge <- rep(1, length(E(g)))
        E(g)$alphaEdge[!glowEdge] <- NA
    } else {
        E(g)$glowEdge <- NA
        E(g)$alphaEdge <- NA
    }

    E(g)$width <- E(g)$color
    edgeWName <- "strength"

    if (sizeDep){
        sizeLab <- "-dependency"
        filteredDep <- filteredDep %>%
                        filter(.data$gene_name %in% names(V(g))) %>%
                        arrange(match(.data$gene_name, names(V(g))))

        # Subset to those dependency scores are available
        if (!is.null(otherVar)) {
            depSubG <- V(g)[names(V(g)) %in% c(filteredDep$gene_name) | 
            names(V(g)) %in% tail(colnames(pcs), n=dim(otherVar)[2])]
        } else {
            depSubG <- V(g)[names(V(g)) %in% c(filteredDep$gene_name)]
        }

        g <- igraph::subgraph(g, depSubG)
        V(g)$size <- vapply(names(V(g)),
            function(x) ifelse(x %in% filteredDep$gene_name,
            -1 * as.numeric(subset(filteredDep,
                filteredDep$gene_name==x)$dependency), NA),
                FUN.VALUE=1)#-1 * filteredDep$dependency
        #meanExp <- apply(pcs[, names(V(g))], 2, mean)
        meanExpCol <- vapply(names(V(g)),
            function(x) ifelse(x %in% geneNames,
            mean(pcs[, x]), mean(apply(pcs[,geneNames], 2, mean))),
            FUN.VALUE=1) # colorize the other nodes later
    } else {
        sizeLab <- "expression"
        #meanExp <- apply(pcs[, names(V(g))], 2, mean)
        meanExpCol <- vapply(names(V(g)),
            function(x) ifelse(x %in% geneNames, mean(pcs[, x]),
            mean(apply(pcs[,geneNames], 2, mean))),
            FUN.VALUE=1) # colorize the other nodes later
        meanExpSize <- meanExpCol
        meanExpSize[is.na(meanExpSize)] <- 2
        V(g)$size <- meanExpSize
    }

    V(g)$color <- meanExpCol

    if (!is.null(hub)){
        ## Make color for glowing nodes
        defHub <- hScore[order(hScore, decreasing=TRUE)][seq_len(hub)]
        nodeShape <- names(V(g)) %in% names(defHub)

        V(g)$cyberColor <- V(g)$color
        V(g)$cyberColor[!nodeShape] <- NA

        V(g)$cyberAlpha <- rep(1, length(V(g)))
        V(g)$cyberAlpha[!nodeShape] <- NA
    } else {
        V(g)$cyberColor <- NA
        V(g)$cyberAlpha <- NA
    }

    V(g)$shape <- rep(19, length(V(g)))

    ## Cluster for multiple pathways
    # if (length(pathNum) > 1) {
    #     V(g)$Pathway <- vapply(names(V(g)),
    #         function(x) ifelse(x %in% geneNames,
    #             cls[x, ]$Pathway, "other variables"), FUN.VALUE="character")
    # }

    ## Plot
    if (length(pathNum) == 1) {
        delG <- delete.vertices(g, igraph::degree(g)==0)

        if (edgeLink) {
            p <- ggraph(delG, layout=layout) +
                geom_edge_link(edge_alpha=0.3,
                    position="identity",
                    aes_(edge_colour=~color, width=~width, label=~label),
                    label_size=3,
                    label_colour=NA,
                    family=fontFamily,
                    angle_calc = "along",
                    label_dodge=unit(3,'mm'),
                    arrow=arrow(length=unit(4, 'mm')),
                    end_cap=circle(5, 'mm'))
        } else {
            p <- ggraph(delG, layout=layout) +
                geom_edge_diagonal(edge_alpha=0.3,
                    position="identity",
                    aes_(edge_colour=~color, width=~width, label=~label),
                    label_size=3,
                    label_colour=NA,
                    family=fontFamily,
                    angle_calc = "along",
                    label_dodge=unit(3,'mm'),
                    arrow=arrow(length=unit(4, 'mm')),
                    end_cap=circle(5, 'mm'))            
        }
        p <- p + geom_node_point(aes_(color=~color, size=~size, shape=~shape),
                            show.legend=TRUE, alpha=0.4)+
            scale_color_continuous(low=nodePal[1], high=nodePal[2],
                name="expression",
                na.value="transparent") +
            scale_size(range=c(scaleSizeLow, scaleSizeHigh) * cexCategory,
                name=sizeLab)+
            scale_edge_width(range=c(1, 3), guide="none")+
            scale_edge_color_continuous(low=edgePal[1],
                high=edgePal[2], name="strength",
                na.value="transparent")+
            guides(edge_color = guide_edge_colorbar(title.vjust = 3)) +
            scale_shape_identity()+
            theme_graph() +
            ggtitle(res[pathNum, "Description"])+
            theme(
                text=element_text(size=16, family=fontFamily, color=textCol),
                plot.title = element_text(size=titleSize, family=fontFamily,
                    color = titleCol),
                panel.background = element_blank(),
                plot.background = element_rect(fill = backCol, colour = NA))

        ## Glowing phase
        layers <- 10
        size <- 1
        edgeSize <- 0.01
        glow_size <- 1.5
        glow_edge_size <- 0.5

        if (!is.null(hub)){
            for (i in seq_len(layers+1)) {
                p <- p + geom_node_point(
                    aes_(
                        color=~cyberColor,
                        alpha=~cyberAlpha
                    ),
                    fill=NA,
                    size=size+(glow_size*i))
            }
        }

        if (!is.null(glowEdgeNum)){
            for (i in seq_len(layers+1)) {
                if (edgeLink) {
                    p <- p + geom_edge_link(
                        position="identity",
                        aes_(edge_colour=~glowEdge,
                            edge_alpha=~alphaEdge),
                        width=edgeSize+(glow_edge_size*i),
                        arrow=arrow(length=unit(i*0.1, 'mm')),
                        end_cap=circle(5, 'mm')
                    )
                } else {
                    p <- p + geom_edge_diagonal(
                        position="identity",
                        aes_(edge_colour=~glowEdge,
                            edge_alpha=~alphaEdge),
                        width=edgeSize+(glow_edge_size*i),
                        arrow=arrow(length=unit(i*0.1, 'mm')),
                        end_cap=circle(5, 'mm')
                    )                    
                }
            }
        }
        bg.colour <- ifelse(is.null(bg.colour), "transparent", bg.colour)
        p <- p + geom_node_text(aes_(label=~name, color=~color),
                check_overlap=TRUE, nudge_y=0.2, repel=TRUE, fontface="bold",
                family=fontFamily, size = labelSize, bg.colour=bg.colour,
                bg.r = bg.r)
        p <- p + scale_alpha(range = c(0.01, 0.1), guide="none")+
            scale_edge_alpha(range=c(0.05, 0.1),guide="none")

    } else if (length(pathNum) > 1) {
        stop("The plotting across multiple pathways is 
            currently not supported in custom.")
    }
    if (strengthPlot){
        p <- p / stp + plot_layout(nrow=2, ncol=1, heights=c(0.8, 0.2))
    }
    if (returnNet){
        returnList <- list()
        returnList[["plot"]] <- p
        returnList[["str"]] <- strength
        returnList[["av"]] <- av
        returnList[["df"]] <- pcs
        return(returnList)
    } else {
        return(p)
    }
}
